### PR TITLE
CI flows on branches use stable releases

### LINKED ISSUE #721: CI flows on branches should use stable releases
Currently CI flows on branches use nightly releases. For example,

https://gubernator.knative.dev/build/knative-prow/logs/ci-knative-eventing-0.5-continuous/1123145177083416576

fails when installing Istio from the latest Knative Serving nightly. The flow should be using artifacts from Knative Serving 0.5.

@mattmoor @tcnghia to validate the assumption.

### PR DESCRIPTION
**Pull request by Fredy-Z**
Fixes #721 

Notes:
1. This change can fix the CI flows for future branches(>=0.6). For fixing the tests failure of [Eventing-0.5](https://testgrid.knative.dev/knative-0.5#eventing), I'm not sure if it would be fine to use the Istio AddOn instead of Istio released by Serving-0.5.
2. We'll need to update the vendor for eventing-sources repo after this PR is merged.

### CODE REVIEW

**Comment by adrcunha** [comment]
Adding an environment variable is unnecessary complexity. As I mentioned in the issue, just check the current git branch.

File: scripts/library.sh [hunk 1]
```diff
@@ -23,7 +23,13 @@ readonly SERVING_GKE_VERSION=gke-latest
 readonly SERVING_GKE_IMAGE=cos
 
 # Public latest stable nightly images and yaml files.
-readonly KNATIVE_BASE_YAML_SOURCE=https://storage.googleapis.com/knative-nightly/@/latest
+KNATIVE_BASE_YAML_SOURCE=""
```
Comment (adrcunha) [inline scripts/library.sh]: Move any code/condition dependent initialization to the end of the file, together with the other vars.

Comment (Fredy-Z) [reply scripts/library.sh]: Done.

File: scripts/library.sh [hunk 2]
```diff
@@ -448,8 +442,34 @@ function get_canonical_path() {
...
+}
+
 # Initializations that depend on previous functions.
 # These MUST come last.
 
 readonly _TEST_INFRA_SCRIPTS_DIR="$(dirname $(get_canonical_path ${BASH_SOURCE[0]}))"
 readonly REPO_NAME_FORMATTED="Knative $(capitalize ${REPO_NAME//-/})"
+
+# Public latest stable nightly images and yaml files.
+readonly BRANCH_VERSION="$(get_branch_version)"
```
Comment (adrcunha) [inline scripts/library.sh]: Thanks, this is way nicer. Once last simplification: you don't need `BRANCH_VERSION` and top-level `if`. `get_branch_version()` can already return the right string to use as `KNATIVE_BASE_YAML_SOURCE`; it even makes sense to rename it to something like `get_knative_base_yaml_source()`.

File: scripts/library.sh [hunk 3]
```diff
@@ -448,8 +442,34 @@ function get_canonical_path() {
   echo "$(cd ${path%/*} && echo $PWD/${path##*/})"
 }
 
+# Return the version for the current branch.
+function get_branch_version() {
+  local branch_name="$(git rev-parse --abbrev-ref HEAD)"
+  # Return the actual version number for the release branch.
+  # For example, we will return "0.5" if the branch name is "release-0.5".
+  if [[ $name =~ ^release-[0-9\.]+ ]]; then
```
Comment (adrcunha) [inline scripts/library.sh]: nit: end the regex with $

File: scripts/library.sh [hunk 4]
```diff
@@ -448,8 +442,34 @@ function get_canonical_path() {
...
+    echo ${branch_name} | sed 's/release-//g'
```
Comment (adrcunha) [inline scripts/library.sh]: I think it's safer if, once in a branch, you find the latest tag and use it; e.g. in branch release-0.5 you would use the tag v0.5.2.

File: scripts/library.sh (hunk 3 above)
Comment (Fredy-Z) [reply scripts/library.sh]: Done.

File: scripts/library.sh (hunk 4 above)
Comment (Fredy-Z) [reply scripts/library.sh]: Sorry I didn't quite get your point. Shouldn't  the yaml files we use here correspond to the code we use to run test? Or do you mean when `Prow` runs the CI jobs for branch release-0.5, it actually uses the code tagged v0.5.2?

File: scripts/library.sh (hunk 2 above)
Comment (Fredy-Z) [reply scripts/library.sh]: Done.

File: scripts/library.sh (hunk 4 above)
Comment (adrcunha) [reply scripts/library.sh]: Yes, they should. Prow runs on the latest commit of a branch, but releases are tagged, so the best approximation is using the latest tag.

Comment (Fredy-Z) [reply scripts/library.sh]: Thanks for the explanation. Done.

File: scripts/library.sh [hunk 5]
```diff
@@ -448,8 +442,27 @@ function get_canonical_path() {
...
 }
 
+# Return the base url we use to build the actual knative yaml sources.
+function get_knative_base_yaml_source() {
+  local knative_base_yaml_source=https://storage.googleapis.com/knative-nightly/@/latest
+  local branch_name="$(git rev-parse --abbrev-ref HEAD)"
+  # If it's a release branch, we should have a different knative_base_yaml_source.
+  if [[ $branch_name =~ ^release-[0-9\.]+$ ]]; then
+    # Get the latest tag name for the current branch, which is likely formatted as v0.5.0
+    local tag_name=$(git describe --tags)
```
Comment (adrcunha) [inline scripts/library.sh]: Better being safe here: sort tags and filter by the release number.

File: scripts/library.sh [hunk 6]
```diff
@@ -448,8 +442,27 @@ function get_canonical_path() {
   echo "$(cd ${path%/*} && echo $PWD/${path##*/})"
 }
 
+# Return the base url we use to build the actual knative yaml sources.
+function get_knative_base_yaml_source() {
+  local knative_base_yaml_source=https://storage.googleapis.com/knative-nightly/@/latest
```
Comment (adrcunha) [inline scripts/library.sh]: nit: quote the strings, please; here and everywhere.

**Pull request merged**